jwtmiddleware.go

```diff
@@ -26,95 +61,132 @@ type TokenExtractor func(r *http.Request) (string, error)
+	// token validated.
+	// Default: true
+	validateOnOptions bool
 }
 
-type JWTMiddleware struct {
-	Options Options
+// Option is how options for the middleware are setup.
+type Option func(*JWTMiddleware)
+
+func WithValidateToken(vt ValidateToken) Option {
+	return func(m *JWTMiddleware) {
+		m.validateToken = vt
+	}
 }
 
-func OnError(w http.ResponseWriter, r *http.Request, err string) {
-	http.Error(w, err, http.StatusUnauthorized)
+func WithContextKey(k string) Option {
+	return func(m *JWTMiddleware) {
+		m.contextKey = k
+	}
 }
 
-// New constructs a new Secure instance with supplied options.
-func New(options ...Options) *JWTMiddleware {
+func WithErrorHandler(h ErrorHandler) Option {
+	return func(m *JWTMiddleware) {
+		m.errorHandler = h
+	}
+}
 
-	var opts Options
-	if len(options) == 0 {
-		opts = Options{}
-	} else {
-		opts = options[0]
+func WithCredentialsOptional(value bool) Option {
+	return func(m *JWTMiddleware) {
+		m.credentialsOptional = value
 	}
+}
 
-	if opts.UserProperty == "" {
-		opts.UserProperty = "user"
+func WithTokenExtractor(e TokenExtractor) Option {
+	return func(m *JWTMiddleware) {
+		m.tokenExtractor = e
 	}
+}
 
-	if opts.ErrorHandler == nil {
-		opts.ErrorHandler = OnError
+func WithDebug(value bool) Option {
+	return func(m *JWTMiddleware) {
+		m.debug = value
 	}
+}
+
+func WithValidateOnOptions(value bool) Option {
+	return func(m *JWTMiddleware) {
+		m.validateOnOptions = value
+	}
+}
 
-	if opts.Extractor == nil {
-		opts.Extractor = FromAuthHeader
+// New constructs a new Secure instance with supplied options.
+func New(opts ...Option) *JWTMiddleware {
+	m := &JWTMiddleware{
+		validateToken:       func(string) (interface{}, error) { panic("not implemented") },
+		contextKey:          "user",
+		errorHandler:        defaultErrorHandler,
+		credentialsOptional: false,
+		tokenExtractor:      FromAuthHeader,
+		debug:               false,
+		validateOnOptions:   true,
 	}
 
-	return &JWTMiddleware{
-		Options: opts,
+	for _, opt := range opts {
+		opt(m)
 	}
+
+	return m
+}
+
+func defaultErrorHandler(w http.ResponseWriter, r *http.Request, err error) {
+	if errors.Is(err, ErrJWTMissing) {
+		http.Error(w, "", http.StatusBadRequest)
+		return
+	} else if errors.Is(err, ErrJWTInvalid) {
+		http.Error(w, "", http.StatusUnauthorized)
+		return
+	}
+
+	http.Error(w, "", http.StatusInternalServerError)
 }
 
 func (m *JWTMiddleware) logf(format string, args ...interface{}) {
-	if m.Options.Debug {
+	if m.debug {
+		// TODO(joncarl): we should allow this logger to be set
 		log.Printf(format, args...)
 	}
 }
 
 // HandlerWithNext is a special implementation for Negroni, but could be used elsewhere.
 func (m *JWTMiddleware) HandlerWithNext(w http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
-	err := m.CheckJWT(w, r)
+	err := m.CheckJWT(r)
 
-	// If there was an error, do not call next.
-	if err == nil && next != nil {
+	if err != nil {
+		m.errorHandler(w, r, err)
```

Comment:
This is a major change. Before the `CheckJWT` func would handle calling the error handler. However why concern `CheckJWT` with an error handler if we don't need to?